lib/utils/verify-signatures.js

```diff
@@ -0,0 +1,460 @@
+    }
+
+    if (this.npm.config.get('json')) {
+      this.appendOutput(this.makeJSON({ invalid, missing }))
+    } else {
+      const timing = `audited ${edges.length} packages in ${Math.floor(Number(elapsed) / 1e9)}s`
+      const verifiedPrefix = verified ? 'verified signatures, ' : ''
+      this.appendOutput(`${verifiedPrefix}${timing}\n`)
+
+      if (missing.length) {
+        const msg = missing.length === 1 ?
+          `package has a ${chalk.bold(chalk.magenta('missing'))} registry signature` :
+          `packages have ${chalk.bold(chalk.magenta('missing'))} registry signatures`
+        this.appendOutput(
+          `${missing.length} ${msg} but the registry is ` +
+          `providing signing keys${this.npm.config.get('missing') ? ':\n' : ''}`
+        )
+        // TODO: This might not be the right option for this
+        if (this.npm.config.get('missing')) {
+          this.appendOutput(this.humanOutput(missing))
+        } else {
+          this.appendOutput(`  run \`npm audit signatures --missing\` for details`)
+        }
+      }
+
+      if (invalid.length) {
+        const msg = invalid.length === 1 ?
+          `package has an ${chalk.bold(chalk.red('invalid'))} registry signature` :
+          `packages have ${chalk.bold(chalk.red('invalid'))} registry signatures`
+        this.appendOutput(
+          `${missing.length ? '\n' : ''}${invalid.length} ${msg}:\n`
+        )
+        this.appendOutput(this.humanOutput(invalid))
+        const plural = invalid.length === 1 ? '' : 's'
+        this.appendOutput(
+          `\nSomeone might have tampered with the package${plural} ` +
+          `since being published on the registry (monster-in-the-middle attack)!\n`
+        )
+      }
+    }
+  }
+
+  findAllRegistryUrls (edges, opts) {
+    return Array.from(edges.reduce((p, edge) => {
+      let alias = false
+      try {
+        alias = npa(edge.spec).subSpec
+      } catch (err) {
+      }
+      const spec = npa(alias ? alias.name : edge.name)
+      p.add(fetch.pickRegistry(spec, opts))
+      return p
+    }, new Set()))
+  }
+
+  appendOutput (...args) {
+    args = [...args].flat()
+    for (const arg of args) {
+      this.output.push(arg)
+    }
+  }
+
+  report () {
+    return { report: this.output.join('\n'), exitCode: this.exitCode }
+  }
+
+  getEdges (nodes, type) {
+    // when no nodes are provided then it should only read direct deps
+    // from the root node and its workspaces direct dependencies
+    if (!nodes) {
+      this.getEdgesOut(this.tree)
+      this.getWorkspacesEdges()
+      return
+    }
+
+    for (const node of nodes) {
+      type === 'edgesOut'
+        ? this.getEdgesOut(node)
+        : this.getEdgesIn(node)
+    }
+  }
+
+  getEdgesIn (node) {
+    for (const edge of node.edgesIn) {
+      this.trackEdge(edge)
+    }
+  }
+
+  getEdgesOut (node) {
+    // TODO: normalize usage of edges and avoid looping through nodes here
+    if (this.npm.config.get('global')) {
```

Comment:
is this command _useful_ to do globally? Many npm commands aren’t. It might be better to error out in global mode for now if there’s not a compelling use case.